driver/src/main/java/org/neo4j/driver/internal/handlers/pulln/AutoPullResponseHandler.java

```diff
@@ -198,11 +217,27 @@ private void enqueueRecord( Record record )
             records = new ArrayDeque<>();
         }
 
+        // too many records in the queue, pause auto request gathering
+        if ( records.size() > HIGH_RECORD_WATERMARK )
+        {
+            isAutoPullEnabled = false;
+        }
+
         records.add( record );
     }
 
     private Record dequeueRecord()
     {
+        if ( records.size() < LOW_RECORD_WATERMARK )
```

Comment:
"<=" for the case when fetchSize = 1. (or any case where the HIGH_RECORD_WATERMARK == LOW_RECORD_WATERMARK)